---
title: "02_visualise"
author: "Alexandra Tidd"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output:
  html_document:
    fig_width: 8
    keep_md: true
    code_folding: show
    toc: true
    toc_float: true
    toc_collapsed: true
    toc_depth: 4
    theme: lumen
---

```{r setup, include = F, message = F, warning = F, class.source = 'fold-hide'}
# module load R/4.4.0 ; Rscript -e "rmarkdown::render('src/02_visualise.Rmd', output_dir = 'out/analysis/')" # nolint: line_length_linter.

# knitr opts
knitr::opts_knit$set(root.dir = "../")
knitr::opts_chunk$set(warning = FALSE, dpi = 300, message = FALSE)

# libraries
library(magrittr)
library(ggplot2)
```

```{r setup2, include = F, message = F, warning = F, class.source = 'fold-hide'}
source("bin/values_to_colors.R")
```

# Load SNVs and indels

We load the matrix of variant counts from the `Sequoia` output. This includes 
all indels and SNVs after filtering germline and artifactual variants.

```{r load_muts}
nv_filtered <-
  read.table("out/PD56034/sequoia/PD56034_both_NV_filtered_all.txt") %>%
  tibble::as_tibble(rownames = "var") %>%
  tidyr::separate_wider_delim(cols = "var", delim = "_",
                              names = c("chr", "pos", "ref", "alt")) %>%
  # classify variant types
  dplyr::mutate(
    var_type = dplyr::case_when(nchar(ref) + nchar(alt) > 2 ~ "indel_count",
                                nchar(ref) + nchar(alt) == 2 ~ "snv_count")) %>%
  # binarise counts
  dplyr::mutate(dplyr::across(where(is.numeric), ~ as.numeric(.x > 0))) %>%
  # get mutation counts
  dplyr::group_by(var_type) %>%
  dplyr::summarise(dplyr::across(where(is.numeric), sum)) %>%
  # get one row per colony
  tidyr::pivot_longer(cols = -var_type, names_to = "colony_id",
                      values_to = "count") %>%
  tidyr::pivot_wider(names_from = var_type, values_from = count) %>%
  # remove normal
  dplyr::filter(colony_id != "normal")
```

# Load colony metadata and CanApps QC

```{r load_metadata}
# load QC from canapps
colony_seq_qc <-
  readxl::read_xls(
    path = "data/Cancer_Pipeline_Reports_SampleDetails.xls",
    skip = 2
  ) %>%
  dplyr::transmute(colony_id = Sample, sample_coverage = `Seq X`)

# load colony metadata, add QC
colony_metadata <-
  readr::read_tsv("out/metadata/colony_metadata.tsv") %>%
  # add coverage
  dplyr::inner_join(colony_seq_qc) %>%
  # get well row and column ids
  tidyr::separate_wider_position(
    cols = "well_id",
    widths = c("well_column_id" = 1, "well_row_id" = 2),
    cols_remove = FALSE
  ) %>%
  # add mutation counts
  dplyr::left_join(nv_filtered) %>%
  # bin coverage
  dplyr::mutate(
    sample_coverage_bin = dplyr::case_when(sample_coverage < 10 ~ "<10x",
                                           sample_coverage > 10 ~ ">10x"))
```

# Plot mutation count and coverage

## Coverage by plate

```{r plot_cov_by_plate_bar}
# colour by timepoint x plate id, ordered by row
colony_metadata %>%
  dplyr::mutate(colony_id = forcats::fct_reorder(colony_id, well_row_id)) %>%
  ggplot(aes(x = colony_id, y = sample_coverage, fill = colony_timepoint)) +
  geom_col() +
  scale_x_discrete(guide = guide_axis(angle = -90)) +
  ggh4x::facet_nested(~ colony_timepoint + plate_id,
    scales = "free_x",
    space = "free_x"
  ) +
  theme_classic() +
  theme(legend.position = "none",
        axis.text.x = element_blank())
```

While the two post-chemotherapy plates show fairly consistent coverage across 
colonies, the pre-chemotherapy plate has a few colonies with very high coverage,
while most colonies have ~ half the coverage of post-chemotherapy.

According to the _Limitations_ section of the 
[Sequoia paper](https://www.nature.com/articles/s41596-024-00962-8):

> The protocol is designed for samples with a high proportion of confident 
> genotyping calls. Sequencing approaches with high rates of artifact or very 
> low coverage may result in a large number of false positives, false negatives 
> or uncertain genotypes, and are unlikely to work well. Good quality sequencing 
> of ≥10x for in vitro single-cell expansions, and ≥20x for LCM samples is 
> generally recommended.

```{r med_cov}
colony_metadata %>%
  dplyr::group_by(colony_timepoint) %>%
  dplyr::summarise(median_coverage = median(sample_coverage),
                   n_samples_over_10x = sum(sample_coverage > 10),
                   n_samples = dplyr::n(),
                   proportion_of_samples_over_10x = n_samples_over_10x /
                     n_samples) %>%
  knitr::kable()
```


Only 22/95 (23%) of the pre-chemotherapy colonies have coverage over 10x. This
is likely to result in a large number of false negatives, making the resultant
phylogeny quite unreliable.

## Mutation count by plate

```{r plot_mutation_count_by_plate_bar}
# colour by timepoint x plate id, ordered by row
colony_metadata %>%
  dplyr::mutate(colony_id = forcats::fct_reorder(colony_id, well_row_id)) %>%
  tidyr::pivot_longer(cols = c("snv_count", "indel_count"),
                      names_to = "var_type", values_to = "count") %>%
  ggplot(aes(x = colony_id, y = count, fill = colony_timepoint)) +
  geom_col() +
  scale_x_discrete(guide = guide_axis(angle = -90)) +
  ggh4x::facet_nested(var_type ~ colony_timepoint + plate_id,
    scales = "free",
    space = "free_x"
  ) +
  theme_classic() +
  theme(legend.position = "none",
        axis.text.x = element_blank())
```

We can see that this coverage pattern is also reflected in the resultant 
mutation counts, both for SNVs and indels.

## Coverage by plate well

```{r plot_cov_by_well_tile}
colony_metadata %>%
  ggplot(aes(x = well_column_id, y = well_row_id, fill = sample_coverage)) +
  geom_tile() +
  ggh4x::facet_nested(~ plate_id + colony_timepoint) +
  theme_classic()
```

Inspecting the coverage plate-x-well indicates that those 10 colonies with very
high coverage in the pre-chemotherapy plate are all located consecutively, in 
wells A01 to B02. This suggests that the strong change in coverage is the 
result of an experimental error.

## Mutation count by plate well

```{r plot_mut_count_by_well_tile}
# snvs
colony_metadata %>%
  ggplot(aes(x = well_column_id, y = well_row_id, fill = snv_count)) +
  geom_tile() +
  ggh4x::facet_nested(~ plate_id + colony_timepoint) +
  theme_classic()
# indels
colony_metadata %>%
  ggplot(aes(x = well_column_id, y = well_row_id, fill = indel_count)) +
  geom_tile() +
  ggh4x::facet_nested(~ plate_id + colony_timepoint) +
  theme_classic()
```

This is again reflected in the SNV and indel counts. This suggests that the 
mutation counts are strongly influenced by this technical variate, rather than a 
true biological difference.

# Plot mutation counts vs coverage

To be sure, we review the relationship between coverage and mutation count in 
SNVs and indels.

```{r plot_mut_count_vs_cov, fid.width = 10, fig.height = 4}
colony_metadata %>%
  tidyr::pivot_longer(cols = c("snv_count", "indel_count"),
                      names_to = "var_type", values_to = "count") %>%
  ggplot(aes(x = sample_coverage, y = count, color = colony_timepoint)) +
  geom_point() +
  facet_wrap(~ var_type, scales = "free_y") +
  theme_classic()
```

There appears to be a strong linear relationship between coverage and mutation 
count in both SNVs and indels. Additionally, most of the post-chemotherapy 
colonies have a higher sample coverage than the pre-chemotherapy colonies, with
a few outlying pre-chemotherapy colonies with very high coverage, and 
proportionately higher mutational mutation counts. This reflects what was previously
observed.

# Plot coverage by chromosome

It may be that the sharp decrease in coverage is due to truncation of the BAMs,
rather than an experimental error. To check this, we check the coverage of each
BAM across all chromosomes in each sample.

```{r plot_check_trunc, fig.height = 50, fig.width = 15}
reads_by_chr <-
  read.table("out/PD56034/sequoia/PD56034_both_NR_filtered_all.txt") %>%
  tibble::as_tibble(rownames = "var") %>%
  tidyr::separate_wider_delim(cols = "var", delim = "_",
                              names = c("chr", "pos", "ref", "alt")) %>%
  readr::type_convert() %>%
  dplyr::mutate(chr = factor(chr,
                             levels = paste0("chr", c(1:22, "X", "Y")))) %>%
  tidyr::pivot_longer(cols = c(-chr, -pos, -ref, -alt)) %>%
  dplyr::group_by(chr, name) %>%
  dplyr::summarise(reads = sum(value))
reads_by_chr %>%
  ggplot(aes(x = chr, y = reads)) +
  geom_col() +
  scale_x_discrete(guide = guide_axis(angle = -90)) +
  ggh4x::facet_nested(sample_coverage_bin + name ~ ., scales = "free_y", 
                      independent = "y") +
  theme(strip.text.y.right = element_text(angle = 0),
        axis.text.y = element_text(size = 1))
```

# Plot trees

The topology and branch length of these trees makes sense in light of these 
findings. The mutation counts in the pre-chemotherapy colonies are much lower 
than would be expected given the donor age, except for a few of the tips. This 
is due to the low coverage.

```{r plot_trees, fig.height = 10}
colony_metadata %>%
  {
    split(., .$donor_id)
  } %>%
  purrr::map2(., names(.), function(donor_metadata, donor_id) {
    files <-
      file.path("out", donor_id, "sequoia") %>%
      list.files(pattern = "\\.tree$", full.names = TRUE)
    if (length(files) > 0) {
      files %>%
        purrr::map(function(file) {
          # load tree
          tree_obj <- ape::read.tree(file)

          # get rownames
          dm <- tibble::column_to_rownames(donor_metadata, "colony_id")

          # get metadata variables with multiple values
          colours <-
            dm[tree_obj$tip.label, ] %>%
            dplyr::select(-well_id, -indel_count, -snv_count) %>%
            purrr::map(~ length(unique(na.omit(.x)))) %>%
            {
              .[. > 1]
            }  %>%
            purrr::map2(., names(.), function(colour_var_n, colour_var_name) {
              # get palette
              colours <- dm[tree_obj$tip.label, colour_var_name]
              colours <- colours[!is.na(colours)]
              if (is.numeric(colours)) {
                legend <- FALSE
                pal <-
                  values_to_colors(colours,
                    min_val = min(colours),
                    max_val = max(colours)
                  )
                pal <- pal %>% purrr::set_names(colours)
              } else {
                legend <- TRUE
                if (colour_var_n == 2) {
                  pal <- c("red", "blue")
                } else {
                  pal <- c(RColorBrewer::brewer.pal(colour_var_n, "Set3"))
                }
                pal <- pal %>% purrr::set_names(unique(colours))
              }
              tree_colours <-
                pal[dm[tree_obj$tip.label, colour_var_name]]

              for (version in c("branch lengths", "equal branch lengths")) {
                tree_obj_v <- tree_obj
                if (version == "equal branch lengths") {
                  tree_obj_v$edge.length <-
                    rep(1, length(tree_obj_v$edge.length))
                }
                # plot
                plot(tree_obj_v,
                  tip.color = tree_colours, cex = 0.5,
                  main = paste0(basename(file), "\n", colour_var_name, ", ",
                                version))
                ape::axisPhylo(side = 1, backward = FALSE)
                if (legend) {
                  legend("topright",
                         legend = unique(names(tree_colours)),
                         col = unique(tree_colours), lty = 1:2, cex = 0.8,
                         title = colour_var_name)
                }
              }
            })
        })
    }
  })
```